### PR TITLE
Registration client for mesos turbo and configuration for mesos target

### PR DESCRIPTION
**Pull request by pallavidn**
Added a Registration client for Mesos Turbo Probe. Also added a target configuration for Mesos Target.

### CODE REVIEW

File: pkg/conf/mesos_target_conf.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+	"io/ioutil"
+	"encoding/json"
+	"github.com/golang/glog"
+	"fmt"
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+	"github.com/turbonomic/turbo-go-sdk/pkg/probe"
+)
+
+// Configuration Parameters to connect to a Mesos Target
+type MesosTargetConf struct {
```
Comment (DongyiYang) [inline pkg/conf/mesos_target_conf.go:13]: Maybe do a nested json, like what we did in [TurboCommunicationConfig](https://github.com/turbonomic/turbo-go-sdk/blob/master/pkg/service/turbo_communication_config.go#L28) in go-sdk.

File: pkg/conf/mesos_target_conf.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+// Create a new MesosTargetConf from a json file.
+// Return null config if the there are errors loading or parsing the file
+func NewMesosTargetConf(targetConfigFilePath string) (*MesosTargetConf, error) {
+	glog.Infof("[MesosClientConf] Target configuration from %s", targetConfigFilePath)
+	config, err := readConfig(targetConfigFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("[[MesosTargetConf] Error reading config "+ targetConfigFilePath + " : " , err)
+	}
+	if config == nil {
```
Comment (DongyiYang) [inline pkg/conf/mesos_target_conf.go]: this check is duplicated with the err check. If err != nil, then config is nil.

File: pkg/conf/mesos_target_conf.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+	if (config.FrameworkIP == "") {
+		config.FrameworkIP = config.MasterIP
+		config.FrameworkPort = config.MasterPort
+		config.Framework = DCOS_Marathon
+	}
+}
+
+// Create a new MesosTargetConf from a given list of AccountValues
+// Return null config if the there are errors validating the config
+func CreateMesosTargetConf(targetType string, accountValues []*proto.AccountValue) (probe.TurboTargetConf, error) {
```
Comment (DongyiYang) [inline pkg/conf/mesos_target_conf.go:81]: why not passed in targetType as a MesosMasterType.

File: pkg/conf/mesos_target_conf.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+	var mesosMasterType MesosMasterType
+	if targetType == string(Apache) {
+		mesosMasterType = Apache
+	} else if targetType == string(DCOS) {
+		mesosMasterType = DCOS
+	} else {
+		glog.Errorf("Unknown Mesos Master Type " , targetType)
+		return nil, fmt.Errorf("Unknown Mesos Master Type " , targetType)
```
Comment (DongyiYang) [inline pkg/conf/mesos_target_conf.go]: in the message of glog.Errorf and fmt.Errorf, the %s is missing

File: pkg/conf/mesos_target_conf.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+		}
+	}
+
+	// Provide framework ip for dcos mesos
+	dcosMesosTargetConf(config)
+	// Validate
+	ok, err := config.validate()
+	if !ok {
+		glog.Errorf("[MesosTargetConf] Invalid config : " + err.Error())
+		return nil, fmt.Errorf("[MesosTargetConf] Invalid config : " + err.Error())
```
Comment (DongyiYang) [inline pkg/conf/mesos_target_conf.go]: same here

File: pkg/conf/mesos_target_conf.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+	}
+	return config, nil
+}
+
+
+// Get the Account Values to create VMTTarget in the turbo server corresponding to this client
+func (mesosConf *MesosTargetConf) GetAccountValues() []*proto.AccountValue {
+	if _, err := mesosConf.validate(); err != nil {
+		glog.Infof("[GetAccountValues] mesos target config is not valid %s", err)
+		return nil
```
Comment (DongyiYang) [inline pkg/conf/mesos_target_conf.go:138]: do we need to return the error as well?

File: pkg/conf/mesos_target_conf.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+func readConfig(path string) (*MesosTargetConf, error) {
+	file, e := ioutil.ReadFile(path)
+	if e != nil {
+		return nil, fmt.Errorf("File error: ", e)
+	}
+	var config MesosTargetConf
+	err := json.Unmarshal(file, &config)
+
+	if err != nil {
+		return nil, fmt.Errorf(string(file) + " \nUnmarshall error : " , err)
```
Comment (DongyiYang) [inline pkg/conf/mesos_target_conf.go]: %s is missing

File: pkg/probe/registration_client.go [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+		Commodity(vCpuType, false).
+		Commodity(vMemType, false).
+		Commodity(vCpuProvisionedType, false).
+		Commodity(vMemProvisionedType, false).
+		Commodity(clusterType, true).
+		ProbeEntityPropertyDef(supplychain.SUPPLY_CHAIN_CONSTANT_IP_ADDRESS,
+			"IP Address where the Container is running").
+		ExternalEntityPropertyDef(supplychain.VM_IP)
+
+	containerVmExternalLink, _ := containerVmExtLinkBuilder.Build()
```
Comment (DongyiYang) [inline pkg/probe/registration_client.go]: there might be no error, but I think we don't want to ignore if there is any.

**Review by DongyiYang — COMMENTED** [review]
Please check glog and fmt, there are some places %s or %v is missing

File: pkg/conf/mesos_target_conf.go (hunk 3 above)
Comment (pallavidn) [reply pkg/conf/mesos_target_conf.go:81]: Will revisit if the targetType parameter should be string or MesosMasterType. This method would be invoked by the generic TurboProbe to create a new discovery client for a target added from the UI.

File: pkg/conf/mesos_target_conf.go (hunk 6 above)
Comment (pallavidn) [reply pkg/conf/mesos_target_conf.go:138]: Will handle this change in the turbo-go-sdk first

**Pull request merged**